checker/jdk/nullness/src/java/net/URL.java

```diff
@@ -307,7 +307,7 @@
      * @see        java.net.URLStreamHandlerFactory#createURLStreamHandler(
      *                  java.lang.String)
      */
-    public URL(String protocol, String host, int port, String file)
+    public URL(String protocol, @Nullable String host, int port, String file)
```

Comment:
Is this string null for a so-called "relative" URL?  Why aren't all the other formal parameter types also `@Nullable`?  I presume they can also be missing for a relative URL.

Reply:
I'm not really sure what Java is going for here. As you say, a relative URL should allow omitting other components, like protocol. My impression from the Javadoc is that an instance of `URL` is intended to represent only absolute URLs (though it does support _resolving_ a relative URL into an absolute URL if you pass the relative URL as a `String spec` to a constructor that accepts a `URL context`).

The only official doc I've found for a null host is in [the Javadoc for `equals`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/net/URL.html#equals(java.lang.Object)):

> Two hosts are considered equivalent if both host names can be resolved into the same IP addresses; else if either host name can't be resolved, the host names must be equal without regard to case; or _both host names equal to null_.

As for why I even made this change:

It looks like Guava passes `null` to this constructor to create `file:` URLs [in `ClassPath`](https://github.com/google/guava/blob/c6f48dced81bc0dbfb8df3cfcc9475d6b9e0f460/guava/src/com/google/common/reflect/ClassPath.java#L483). (We've also copied that code to a few locations in our tests.)

I can check whether an empty-string host would work just as well. It might, given that the _parsing_ constructor, `new URL(String)`, turns an input of `file:///home/cpovirk` into an object with a empty-string host, not a null host.

Reply:
It does look like we can pass an empty string just as well as `null`. That suggests that accepting `null` here isn't essential. Let me know if you'd like me to revert this part.

Reply:
This seems OK to me either way; it's not worth delaying the pull request over, and null is a better value to use than the empty string in any event.  Thank you for investigating!